File: src/emulator/peripherals/eeprom/eeprom.ts
import { BoardComponent } from "../../board/boardComponent.ts";
import {Memory} from "../../memory/memory.ts";
import {Board} from "../../board/board.ts";

export const eepromCommands = {
    WRITE_ENABLE: 0b0000_0110,
    WRITE_DISABLE: 0b0000_0100,
    READ_STATUS: 0b0000_0101,
    WRITE_STATUS: 0b0000_0001,
    READ_MEMORY: 0b0000_0011,
    WRITE_MEMORY: 0b0000_0010
}

export const eepromFlags = {
    status: {
        WRITE_UNLOCK: 0x2
    }
}


export const eepromState = {
    waiting: 0,
    gettingStatus: 1,
    gettingAddressHigh: 2,
    gettingAddressLow: 3,
    gettingBytes: 4
}

export class EepRom extends BoardComponent {
    memory: Memory
    state: number = eepromState.waiting

    status: number = 0

    highAddress: number = 0
    lowAddress: number = 0
    offset: number = 0

    constructor(board: Board, eepromMemory: Memory) {
        super(board)
        this.memory = eepromMemory
    }


}

File: PocketWalker/Emulator/Peripherals/Eeprom/Eeprom.cpp
#include "Eeprom.h"

#include "../../Ssu/Ssu.h"

namespace EepromFlags
{
    enum Commands : uint8_t
    {
        WRITE_ENABLE = 0b00000110,
        WRITE_DISABLE = 0b00000100,
        READ_STATUS = 0b00000101,
        WRITE_STATUS = 0b00000001,
        READ_MEMORY = 0b00000011,
        WRITE_MEMORY = 0b00000010
    };

    enum Status : uint8_t
    {
        WRITE_UNLOCK = 0b00000010
    };
}

void Eeprom::TransmitAndReceive(Ssu* ssu)
{
    switch (state)
    {
    case Waiting:
        {
            if (ssu->transmit == EepromFlags::Commands::READ_MEMORY)
            {
                state = GettingHighAddress;
            }
            else if (ssu->transmit == EepromFlags::Commands::READ_STATUS)
            {
                state = GettingStatus;
            }
            break;
        }
    case GettingStatus:
        {
            ssu->receive = status;
            ssu->status |= SsuFlags::Status::TRANSMIT_END;
            break;
        }
    case GettingHighAddress:
        {
            highAddress = ssu->transmit;
            state = GettingLowAddress;
            break;
        }
    case GettingLowAddress:
        {
            lowAddress = ssu->transmit;
            state = GettingBytes;
            break;
        }
    case GettingBytes:
        {
            ssu->receive = memory->ReadByte((highAddress << 8 | lowAddress) + offset);
            offset++;
            
            ssu->status |= SsuFlags::Status::TRANSMIT_END;
            break;
        }
    }

    ssu->status |= SsuFlags::Status::RECEIVE_FULL;
    ssu->status |= SsuFlags::Status::TRANSMIT_EMPTY;
}

void Eeprom::Transmit(Ssu* ssu)
{
    switch (state)
    {
    case Waiting:
        {
            if (ssu->transmit == EepromFlags::Commands::WRITE_ENABLE)
            {
                status |= EepromFlags::Status::WRITE_UNLOCK;
                ssu->status |= SsuFlags::Status::TRANSMIT_END;
            }
            else if (ssu->transmit == EepromFlags::Commands::READ_MEMORY)
            {
                state = GettingHighAddress;
            }
            break;
        }
    case GettingHighAddress:
        {
            highAddress = ssu->transmit;
            state = GettingLowAddress;
            break;
        }
    case GettingLowAddress:
        {
            lowAddress = ssu->transmit;
            state = GettingBytes;
            break;
        }
    case GettingBytes:
        {
            memory->WriteByte((highAddress << 8 | lowAddress) + offset, ssu->transmit);
            offset++;
            offset %= 128;
            
            ssu->status |= SsuFlags::Status::TRANSMIT_END;
            break;
        }
    }

    ssu->status |= SsuFlags::Status::TRANSMIT_EMPTY;
}

void Eeprom::Reset()
{
    state = Waiting;
    offset = 0;
}
